File: CoreMerge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace LVLTool
{
    /// <summary>
    /// Gathers strings from core.lvl files and merges them into a given 'base' core.lvl file
    /// </summary>
    public class CoreMerge
    {
        UcfbHelper mCoreUucfbFileHelper = null;
        
        // 'mStringsToAdd' will be used in the 'Gather' operation
        // key = lang, val = content
        Dictionary<string, StringBuilder> mStringsToAdd = new Dictionary<string, StringBuilder>();

        /// <summary>
        /// The base core.lvl file, typically should be named 'base.core.lvl'
        /// </summary>
        public CoreMerge(string baseCoreFile)
        {
            mCoreUucfbFileHelper = new UcfbHelper();
            mCoreUucfbFileHelper.FileName = baseCoreFile;
            mCoreUucfbFileHelper.InitializeRead();
        }

        /// <summary>
        /// Gather the strings from a core.lvl file or .txt file in the correct string format.
        /// </summary>
        /// <param name="file"></param>
        public void GatherStrings(string file)
        {
            if (file.EndsWith(".lvl", StringComparison.InvariantCultureIgnoreCase))
            {
                UcfbHelper ucfbFileHelper = new UcfbHelper();
                ucfbFileHelper.FileName = file;
                ucfbFileHelper.InitializeRead();

                Chunk cur = null;
                while ((cur = ucfbFileHelper.RipChunk(false)) != null)
                {
                    if (cur.Type == "Locl")
                    {
                        cur.LocHelper = new LocHelper(cur.GetAssetData());
                        string allStrings = cur.LocHelper.GetAllStrings();
                        if (mStringsToAdd.ContainsKey(cur.Name))
                        {
                            StringBuilder sb = mStringsToAdd[cur.Name];
                            if (sb[sb.Length - 1] != '\n')
                                sb.Append("\n");
                            sb.Append(allStrings);
                        }
                        else
                        {
                            mStringsToAdd.Add(cur.Name, new StringBuilder(allStrings));
                        }
                    }
                }
            }
            else if (file.EndsWith(".txt", StringComparison.InvariantCultureIgnoreCase))
            {
                int slashLoc = file.LastIndexOf('\\');
                string langName = "";
                if (slashLoc > -1)
                {
                    langName = file.Substring(slashLoc + 1).Replace(".txt", "");
                    string allStrings = File.ReadAllText(file);
                    if (mStringsToAdd.ContainsKey(langName))
                    {
                        StringBuilder sb = mStringsToAdd[langName];
                        if (sb[sb.Length - 1] != '\n')
                            sb.Append("\n");
                        sb.Append(allStrings);
                    }
                    else
                    {
                        mStringsToAdd.Add(langName, new StringBuilder(allStrings));
                    }
                }
            }
            else
            {
                Console.WriteLine("skipping file '{0}'", file);
            }
        }

        private void AddGatheredStrings()
        {
            if (mStringsToAdd.Count > 0)
            {
                Chunk cur = null;
                List<Chunk> locChunks = new List<Chunk>();
                // rip through the whole thing first.
                while ((cur = mCoreUucfbFileHelper.RipChunk(false)) != null)
                {
                    if (cur.Type == "Locl") locChunks.Add(cur);
                }

                Chunk c = null;
                // process from back to front so we don't mess up the chunk's knowledge of it's position
                for(int i = locChunks.Count-1; i> -1; i--)
                {
                    c = locChunks[i];
                    if (mStringsToAdd.ContainsKey(c.Name))
                    {
                        LocHelper helper = new LocHelper(c.GetAssetData());
                        helper.AddNewStrings(mStringsToAdd[c.Name].ToString() );
                        mCoreUucfbFileHelper.ReplaceUcfbChunk(c, helper.GetUcfbData(), true);
                    }
                }
            }
        }

        /// <summary>
        /// Save to the 
        /// </summary>
        /// <param name="fileName"></param>
        public void Save(string fileName)
        {
            AddGatheredStrings();
            mCoreUucfbFileHelper.SaveData(fileName);
        }

        public static void MergeLoc(string baseFileName, string saveFileName, List<string> files)
        {
            CoreMerge cm = new CoreMerge(baseFileName);
            foreach (string file in files)
            {
                Console.WriteLine("Gathering strings from '{0}'", file);
                try
                {
                    cm.GatherStrings(file);
                }
                catch (Exception ex)
                {
                    Program.MessageUser(String.Format("Exception encountered while processing file {0}\n{1}\n{2}", 
                        file, 
                        ex.Message,
                        ex.StackTrace ));
                }
            }
            cm.Save(saveFileName);
        }

    }
}

File: dart_translation/lib/core_merge.dart
/// Merges localization strings from core.lvl files into a base core.lvl.
/// Translated from C# CoreCode/CoreMerge.cs for Linux/CLI use.

import 'dart:io';

import 'loc_helper.dart';
import 'ucfb_helper.dart';
import 'package:dart_translation/lvltool_config.dart';

class CoreMerge {
  UcfbHelper? _coreUcfbFileHelper;
  final Map<String, StringBuffer> _stringsToAdd = {};

  CoreMerge(String baseCoreFile) {
    _coreUcfbFileHelper = UcfbHelper();
    _coreUcfbFileHelper!.fileName = baseCoreFile;
    _coreUcfbFileHelper!.initializeRead();
  }

  void gatherStrings(String file) {
    if (file.toLowerCase().endsWith('.lvl')) {
      final ucfbFileHelper = UcfbHelper();
      ucfbFileHelper.fileName = file;
      ucfbFileHelper.initializeRead();
      Chunk? cur;
      while ((cur = ucfbFileHelper.ripChunk(false)) != null) {
        if (cur != null &&  cur.type == 'Locl') {
          cur.locHelper = LocHelper(cur.getAssetData());
          final allStrings = cur.locHelper!.getAllStrings();
          if (_stringsToAdd.containsKey(cur.name)) {
            final sb = _stringsToAdd[cur.name]!;
            final s = sb.toString();
            if (s.isNotEmpty && s.codeUnitAt(s.length - 1) != 10) {
              sb.writeln();
            }
            sb.write(allStrings);
          } else {
            _stringsToAdd[cur.name] = StringBuffer(allStrings);
          }
        }
      }
    } else if (file.toLowerCase().endsWith('.txt')) {
      final sep = file.contains('/') ? '/' : '\\';
      final lastSlash = file.lastIndexOf(sep);
      if (lastSlash > -1) {
        final langName = file.substring(lastSlash + 1).replaceAll('.txt', '');
        final allStrings = File(file).readAsStringSync();
        if (_stringsToAdd.containsKey(langName)) {
          final sb = _stringsToAdd[langName]!;
          final s = sb.toString();
          if (s.isNotEmpty && s.codeUnitAt(s.length - 1) != 10) {
            sb.writeln();
          }
          sb.write(allStrings);
        } else {
          _stringsToAdd[langName] = StringBuffer(allStrings);
        }
      }
    } else {
      print("info: skipping file '$file'");
    }
  }

  String? getStrings(String language) {
    return _stringsToAdd[language]?.toString();
  }

  String? getAllStrings() {
    if (_stringsToAdd.isEmpty) return null;
    final retVal = StringBuffer();
    for (String key in _stringsToAdd.keys) {
      retVal.writeln('//Language: $key');
      retVal.write(_stringsToAdd[key].toString());
    }
    return retVal.toString();
  }

  void _addGatheredStrings() {
    if (_stringsToAdd.isEmpty) return;
    Chunk? cur;
    final locChunks = <Chunk>[];
    while ((cur = _coreUcfbFileHelper!.ripChunk(false)) != null) {
      if ( cur != null && cur.type == 'Locl') locChunks.add(cur);
    }
    for (int i = locChunks.length - 1; i > -1; i--) {
      final c = locChunks[i];
      if (_stringsToAdd.containsKey(c.name)) {
        final helper = LocHelper(c.getAssetData());
        helper.addNewStrings(_stringsToAdd[c.name].toString());
        _coreUcfbFileHelper!.replaceUcfbChunkBytes(c, helper.getUcfbData(), true);
      }
    }
  }

  void save(String fileName) {
    _addGatheredStrings();
    _coreUcfbFileHelper!.saveData(fileName);
  }

  static void mergeLoc(String baseFileName, String saveFileName, List<String> files) {
    final cm = CoreMerge(baseFileName);
    for (String file in files) {
      print("info: Gathering strings from '$file'");
      try {
        cm.gatherStrings(file);
      } catch (ex, stack) {
        if (messageUser != null) {
          messageUser!('Exception encountered while processing file $file\n$ex\n$stack');
        }
      }
    }
    cm.save(saveFileName);
  }
}
